scripts-foundry/common/DeployMagicDropTokenImplRegistry.s.sol

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Script} from "forge-std/Script.sol";
+import {MagicDropTokenImplRegistry} from "contracts/registry/MagicDropTokenImplRegistry.sol";
+
+contract DeployMagicDropTokenImplRegistry is Script {
+    error AddressMismatch();
+    
+    function run() external {
+        bytes32 salt = vm.envBytes32("REGISTRY_SALT");
+        address expectedAddress = address(uint160(vm.envUint("REGISTRY_EXPECTED_ADDRESS")));
+        address initialOwner = address(uint160(vm.envUint("INITIAL_OWNER")));
+        uint256 privateKey = vm.envUint("PRIVATE_KEY");
+
+        vm.startBroadcast(privateKey);
+
+        // Deploy the contract using CREATE2 directly
+        MagicDropTokenImplRegistry registryImpl = new MagicDropTokenImplRegistry{salt: salt}();
+
+        // Verify the deployed address matches the predicted address
+        if (address(registryImpl) != expectedAddress) {
+            revert AddressMismatch();
+        }
+
+        registryImpl.initialize(initialOwner);
```

Comment:
`initialize` should be coupled with the deployment transaction. 

Initialize is wide open and can only be called once. I'm used to seeing the initialization params encoded and passed to the deployment function so that after deployment the initialize function is called in the same transaction. 

This prevents outside actors from calling initialize in between the transactions

Reply:
any examples you can share?

I believe with foundry these two calls are done in one transaction

Reply:
Ah I took for granted how Foundry does deploy scripts.

So it looks like the transactions are dependent on Foundry's CREATE2 deployer (0x4e59b44847b379578588920cA78FbF26c0B4956C) so that would make sense that the data could be encoded/called as part of the init bytecode. 

It would be important to note that if we are targeting uniform contract addresses across chains then we are dependent on their CREATE2 contract. I think this is a trade off worth thinking about. 

Here are some `upgradeToAndCall` examples:
[Reference Implementation](https://eips.ethereum.org/EIPS/eip-1967)
[Aave](https://github.com/search?q=repo%3Aaave%2Faave-v3-core%20upgradeToAndCall&type=code)
[Open Zeppelin](https://github.com/search?q=repo%3AOpenZeppelin%2Fopenzeppelin-contracts%20upgradeToAndCall&type=code)

I usually lean on OZ unit tests for implementation details.

Reply:
I think from our convo offline relying on Foundry should be fine. If we ever run into issues with the CREATE2 dependency we can major version bump and pick the convo back up from that point -> e.g. maintaining our own factory deployer. I think the effort v return right now is low.